### DISCUSSION TITLE
How can I set the range for a graph in manim?

### DISCUSSION THREAD

**Discussion by baljeetrathi** [3b1b/manim#1402] [Q&A] [answered]
I have tried two different approaches:

```
top = axes.get_graph(
            lambda x: math.sqrt(1 - (abs(x) - 1)*(abs(x) - 1)),
            x_range =  [-2, 2],
            color = BLUE
        )
```

and 

```
top = axes.get_graph(
            lambda x: math.sqrt(1 - (abs(x) - 1)*(abs(x) - 1)),
            x_range =  range(-2, 2),
            color = BLUE
        )
```

Both of them give me domain error. My axes looks like this:

```
axes = Axes((-5, 5), (-5, 5))
```

How can I show a plot from -2 to 2 while the graph goes from -5 to 5? The graph also does not seem very smooth.

Can anyone please help?

Thanks.

**Comment by Just-Math** [comment]
I can try to use `x_min = -2` and `x_max = 2`, `y_min = -5`, `y_max - 5`

**Reply by baljeetrathi** [reply]
Thanks @Just-Math :)

However, this does not work in newer versions of manim.

**Reply by Just-Math** [reply]
> Thanks @Just-Math :)
> 
> However, this does not work in newer versions of manim.

Maybe `"x_min" : -2` ?

**Reply by baljeetrathi** [reply]
The problem is somewhere within Manim itself. It gives Math domain error for this particular function even though values stay within limits.

**Reply by Just-Math** [reply]
> The problem is somewhere within Manim itself. It gives Math domain error for this particular function even though values stay within limits.

Of course! Because sqrt( (|x| - 1) * (1 - |x|) ) = sqrt( - (1-|x|)^2 ), thus -(1-|x|)^2 < 0. You need to use complex plane

**Comment by ghost** [comment, marked as answer]
@baljeetrathi This code works for me though?
```py
class AxesGraph(Scene):
    def construct(self):
        axes = Axes([-5, 5], [-5, 5])
        graph = axes.get_graph(
            lambda x: math.sqrt(1 - (abs(x) - 1) ** 2),
            x_range=[-2, 2, 0.001]  # x_min, x_max, x_step
        )
        self.add(axes, graph)
```


> The graph also does not seem very smooth

It can be made smooth by decreasing the sample frequency (i.e `x_step`, which is 0.001 here)

**Reply by baljeetrathi** [reply]
Thanks @abhi-6988. The problem is that I keep getting the Math domain error with the exact same values that you have. Is this due to some floating-point error?

**Reply by ghost** [reply]
@baljeetrathi  Oh that's weird, sorry I've no idea on why it shows a domain error for you. Setting `x_range = [-1.9999, 1.9999, 0.001]` is worth giving a try if the offset is not much of an issue.

**Reply by baljeetrathi** [reply]
I ended up using `abs()` to make values positive but your solution is much better. 😀
